x/poolmanager/client/query_proto_wrap.go

```diff
@@ -254,3 +258,378 @@ func (q Querier) TotalLiquidity(ctx sdk.Context, req queryproto.TotalLiquidityRe
 		Liquidity: totalLiquidity,
 	}, nil
 }
+
+// EstimateTradeBasedOnPriceImpact returns the input and output amount of coins for a pool trade
+// based on external price and maximum price impact.
+func (q Querier) EstimateTradeBasedOnPriceImpact(
+	ctx sdk.Context,
+	req queryproto.EstimateTradeBasedOnPriceImpactRequest,
+) (*queryproto.EstimateTradeBasedOnPriceImpactResponse, error) {
+	if req.PoolId == 0 {
+		return nil, status.Error(codes.InvalidArgument, "Invalid Pool Id")
+	}
+
+	if req.FromCoin.Denom == "" {
+		return nil, status.Error(codes.InvalidArgument, "invalid from coin denom")
+	}
+
+	if req.ToCoinDenom == "" {
+		return nil, status.Error(codes.InvalidArgument, "invalid to coin denom")
+	}
+
```

Comment:
Since I'm calling `swapModule.CalculateSpotPrice` directly instead of using the router.go func this logic would need to be moved down to each individual `SpotPrice` calculation, in other queries this check is done before that:
```
// SpotPrice returns target pool asset prices on base and quote assets.
// nolint: staticcheck
func (q Querier) SpotPrice(ctx context.Context, req *types.QuerySpotPriceRequest) (*types.QuerySpotPriceResponse, error) {
	if req == nil {
		return nil, status.Error(codes.InvalidArgument, "empty request")
	}

	if req.BaseAssetDenom == "" {
		return nil, status.Error(codes.InvalidArgument, "invalid base asset denom")
	}

	if req.QuoteAssetDenom == "" {
		return nil, status.Error(codes.InvalidArgument, "invalid quote asset denom")
	}

	sdkCtx := sdk.UnwrapSDKContext(ctx)

	// Note: the base and quote asset argument order is intentionally incorrect
	// due to a historic bug in the original implementation.
	sp, err := q.Keeper.CalculateSpotPrice(sdkCtx, req.PoolId, req.BaseAssetDenom, req.QuoteAssetDenom)
	if err != nil {
		return nil, status.Error(codes.Internal, err.Error())
	}

	return &types.QuerySpotPriceResponse{
		SpotPrice: sp.String(),
	}, nil
}
```
I would suggest leaving as is as I don't know what else relies on the `CalculateSpotPrice` func

Reply:
Oh I was referring to grouping all the logic in this method and moving it to router.go. I felt having a heavy logic in query_proto_wrap a little out of place. Dependency on CalculateSpotPrice should not matter!

Reply:
Done